### PR TITLE
support empty password for mysql

### PR DESCRIPTION
**Pull request by chungngoops**
related issue: https://github.com/testcontainers/testcontainers-java/issues/896

added MYSQL_ALLOW_EMPTY_PASSWORD=yes option when the passwrod is empty.
- [x] verify all tests are passing.

### CODE REVIEW

**Comment by kiview** [comment]
Looks good, can you please add a test case for using an empty password?

**Comment by chungngoops** [comment]
@kiview: sure!

File: modules/jdbc-test/src/test/java/org/testcontainers/junit/EmptyPasswordMysqlTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    private static final String USER = "root"; // MySQL allows empty password with root only
+
+    // Add MYSQL_ROOT_HOST environment so that we can root login from anywhere for testing purposes
+    @Rule
+    public MySQLContainer mysql = (MySQLContainer) new MySQLContainer("mysql:5.5")
+            .withDatabaseName(DB_NAME)
+            .withUsername(USER)
+            .withPassword("")
+            .withEnv("MYSQL_ROOT_HOST", "%");
+
```
Comment (kiview) [inline modules/jdbc-test/src/test/java/org/testcontainers/junit/EmptyPasswordMysqlTest.java]: Just so I can get a better understanding, if MySQL would only allow using empty password with user 'root', wouldn't it be necessary to always set with environment variable if the use an empty password?

Also would it make sense to throw an exception in `configure()` if we have an empty password with non root user combination, so we can fail early and give user feedback?

Comment (chungngoops) [reply modules/jdbc-test/src/test/java/org/testcontainers/junit/EmptyPasswordMysqlTest.java]: Actually, it should be: docker would only allow using an empty password with the user 'root' by using the `MYSQL_ALLOW_EMPTY_PASSWORD` option, otherwise it always come in conjunction `MYSQL_USER, MYSQL_PASSWORD` to create a new `user`.

Yes, throwing an exception gives a good feedback and save time. Do you have any suggestion which exception should be thrown?

Comment (kiview) [reply modules/jdbc-test/src/test/java/org/testcontainers/junit/EmptyPasswordMysqlTest.java]: I think we can just use `ContainerLaunchException`. It's normally used if the wait check fails, but it's a `RuntimeException` and should be enough with a good error message.

File: modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java [hunk 2]
```diff
@@ -41,8 +41,16 @@ protected void configure() {
...
-        addEnv("MYSQL_PASSWORD", password);
-        addEnv("MYSQL_ROOT_PASSWORD", password);
+        if(password != null && !password.isEmpty()){
+          addEnv("MYSQL_PASSWORD", password);
+          addEnv("MYSQL_ROOT_PASSWORD", password);
+        } else {
+          if("root".equalsIgnoreCase(username)){
+            addEnv("MYSQL_ALLOW_EMPTY_PASSWORD", "yes");
+          } else {
+            throw new RuntimeException("Empty password can be used only with the root user");
```
[the remark this answers is not in the document]
Comment (kiview) [reply modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: Yes, please use `ContainerLaunchException`, maybe you misunderstood my comment before.

File: modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java [hunk 3]
```diff
@@ -41,8 +41,16 @@ protected void configure() {
         addExposedPort(MYSQL_PORT);
         addEnv("MYSQL_DATABASE", databaseName);
         addEnv("MYSQL_USER", username);
-        addEnv("MYSQL_PASSWORD", password);
-        addEnv("MYSQL_ROOT_PASSWORD", password);
+        if(password != null && !password.isEmpty()){
+          addEnv("MYSQL_PASSWORD", password);
+          addEnv("MYSQL_ROOT_PASSWORD", password);
+        } else {
+          if("root".equalsIgnoreCase(username)){
```
Comment (kiview) [inline modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: Please use `username` as the object for the method invocation.

File: modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java [hunk 4]
```diff
@@ -41,8 +41,16 @@ protected void configure() {
         addExposedPort(MYSQL_PORT);
         addEnv("MYSQL_DATABASE", databaseName);
         addEnv("MYSQL_USER", username);
-        addEnv("MYSQL_PASSWORD", password);
-        addEnv("MYSQL_ROOT_PASSWORD", password);
+        if(password != null && !password.isEmpty()){
+          addEnv("MYSQL_PASSWORD", password);
+          addEnv("MYSQL_ROOT_PASSWORD", password);
+        } else {
```
Comment (kiview) [inline modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: We can use an `else if` here and lose one level of nesting.

File: modules/jdbc-test/src/test/java/org/testcontainers/junit/SimpleMySQLTest.java [hunk 5]
```diff
@@ -128,6 +133,20 @@ public void testMySQL8() throws SQLException {
...
+    public void testEmptyPasswordWithNonRootUser() {
+      thrown.expect(RuntimeException.class);
+
+      MySQLContainer container = (MySQLContainer) new MySQLContainer("mysql:5.5")
+      .withDatabaseName("TEST")
+      .withUsername("test")
+      .withPassword("")
+      .withEnv("MYSQL_ROOT_HOST", "%");
+
+      container.start();
```
Comment (kiview) [inline modules/jdbc-test/src/test/java/org/testcontainers/junit/SimpleMySQLTest.java]: Let's do a try-catch here, catching `ContainerException`.

File: modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java [hunk 6]
```diff
@@ -36,13 +37,20 @@ public MySQLContainer(String dockerImageName) {
...
         addEnv("MYSQL_USER", username);
-        addEnv("MYSQL_PASSWORD", password);
-        addEnv("MYSQL_ROOT_PASSWORD", password);
+        if (password != null && !password.isEmpty()) {
+            addEnv("MYSQL_PASSWORD", password);
+            addEnv("MYSQL_ROOT_PASSWORD", password);
+        } else if (MYSQL_ROOT_USER.equalsIgnoreCase(username)) {
+            addEnv("MYSQL_ALLOW_EMPTY_PASSWORD", "yes");
+        } else {
+            throw new ContainerLaunchException("Empty password can be used only with the root user");
```
Comment (kiview) [inline modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: I just found out, that we catch all exceptions and throw `ContainerLaunchException` in `GenericContainer`'s `doStart()` method, so it's really not that important which exception we throw here, but `ContainerLaunchException` makes sense of course.

**Comment by kiview** [comment]
LGTM
You want to give it a final look @rnorth?

File: modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java (hunk 6 above)
Comment (chungngoops) [reply modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: @kiview: should we check the cause of the exception to make sure that the ContainerLaunchException is thrown by this method  instead of others

Comment (kiview) [reply modules/mysql/src/main/java/org/testcontainers/containers/MySQLContainer.java]: I was thinking something like this at first also. But in the end, we just want to check the public API and the container behaves as expected if we receive this exception, so I'd be good with it.

**Comment by kiview** [comment]
Merged, thanks a lot @chungngoops.

**Comment by rui-ferreira** [comment]
Unfortunately, this doesn't work if you launch the container via JDBC URL scheme, for example, with `jdbc:mysql:5.7.22://localhost:3306/databasename?user=root&password=`.

The problem is that `JdbcDatabaseContainerProvider` gets the password from the connection URL and instantiates `MySQLContainer` with a default `test` password in case the value in the URL is empty.
This is the problematic line:
`final String password = connectionUrl.getQueryParameters().getOrDefault(pwdParamName, "test");`

Can you provide a fix for this case?

**Comment by rui-ferreira** [comment]
Never mind, just noticed this got fixed in #2207. :)

**Comment by aspodin** [comment]
@Override
    protected void configure() {
        optionallyMapResourceParameterAsVolume(MY_CNF_CONFIG_OVERRIDE_PARAM_NAME, "/etc/mysql/conf.d",
            "mysql-default-conf");

        addEnv("MYSQL_DATABASE", databaseName);
        if (!MYSQL_ROOT_USER.equalsIgnoreCase(username)) {
            addEnv("MYSQL_USER", username);
        }
        if (password != null && !password.isEmpty()) {
            addEnv("MYSQL_PASSWORD", password);
            addEnv("MYSQL_ROOT_PASSWORD", password);
        } else if (MYSQL_ROOT_USER.equalsIgnoreCase(username)) {
            addEnv("MYSQL_ALLOW_EMPTY_PASSWORD", "yes");
        } else {
            throw new ContainerLaunchException("Empty password can be used only with the root user");
        }
        setStartupAttempts(3);
    }

addEnv("MYSQL_ALLOW_EMPTY_PASSWORD", "yes"); will be done for such configuration: 
container = new CustomMySqlContainer(IMAGE_VERSION)
                        .withUsername("root")
                        .withPassword("")
                        .withDatabaseName("users")
                        .withUrlParam("createDatabaseIfNotExist", "true");

but with this configuration container does not start with error message:
 MYSQL_USER="root", MYSQL_USER and MYSQL_PASSWORD are for configuring a regular user and cannot be used for the root user
    Remove MYSQL_USER="root" and use one of the following to control the root user password:
    - MYSQL_ROOT_PASSWORD
    - MYSQL_ALLOW_EMPTY_PASSWORD
    - MYSQL_RANDOM_ROOT_PASSWORD
